projects/extension/src/background/AppMediator.ts

```diff
@@ -117,187 +123,70 @@ export class AppMediator extends (EventEmitter as { new(): StateEmitter }) {
-   * @returns true if this app handled the message otherwise false
-   */
-  processSmoldotMessage(message: JsonRpcResponse): boolean {
-    if (this.#state === 'disconnected') {
-      // Shouldn't happen - we remove the AppMediator from the smoldot's apps
-      // when we disconnect (below).
-      console.error(`Asked a disconnected UApp (${this.name}) to process a message from ${this.#chainName as string}`);
-      return false;
-    }
-
-    if (this.#state === 'disconnecting') {
-      // Handle responses to our unsubscription messages
-      const request = this.requests.find(r => r.chainID === message.id);
-      if (request !== undefined) {
-        // We don't forward the RPC message to the UApp - it's not there any more
-        const idx = this.requests.indexOf(request);
-        this.requests.splice(idx, 1);
-        this.#checkForDisconnected();
-        return true;
-      }
-    }
-
-    // subscription message
-    if (message.method) {
-      if(!(message as JsonRpcResponseSubscription).params?.subscription) {
-        throw new Error('Got a subscription message without a subscription id');
-      }
+  #handleSpecMessage = (msg: MessageToManager, chainName: string): void => {
+    const chainSpec: string = relayChains.has(chainName) ?
+      (relayChains.get(chainName) || '') : msg.payload;
 
-      const sub = this.subscriptions.find(s => s.subID == message.params?.subscription);
-      if (!sub) {
-        // not our subscription
-        return false;
-      }
-
-      this.#port.postMessage({ type: 'rpc', payload: JSON.stringify(message) });
-      return true;
-    }
-
-    // regular message
-    const request = this.requests.find(r => r.chainID === message.id);
-    if (request === undefined) {
-      // Not our message
-      return false;
-    }
-
-    // let's process this message - it's for us
-    const idx = this.requests.indexOf(request);
-    this.requests.splice(idx, 1);
-
-    // is this a response telling us the subID for a subscription?
-    const sub = this.subscriptions.find(s => s.appIDForRequest == request.appID);
-    if (sub) {
-      if (sub.subID) {
-        throw new Error('Found a subscription for this request ID but it already had a sub id');
-      }
-
-      if (!message.result) {
-        throw new Error('Got a message which we expected to return us a subid but it wasnt there');
-      }
-
-      sub.subID = message.result as (string | number | undefined);
+    const rpcCallback = (rpc: string) => {
+      this.#port.postMessage({ type: 'rpc', payload: rpc })
     }
 
-    // change the message ID to the ID the app is expecting
-    message.id = request.appID
-    this.#port.postMessage({ type: 'rpc', payload: JSON.stringify(message) });
-
-    return true;
-  }
-
-  #addChain = async (chainName: string, chainSpecs: string): Promise<void> => {
-    this.#chain = await this.#manager.addChain(chainName, chainSpecs);
+    this.#manager.addChain(chainName, chainSpec, rpcCallback)
```

Comment:
Im not sure why this is communicated first to the manager and then added to the local var.

In addition as discussed (but did not ended up with a decision) - there should not be pending messages arriving until the chain is connected. Am I missing something? If for any reason this does not conclude user will not be aware that the pending messages were not delivered.

The flow below does not guarantee the deliverance of the messages (for eample `if (!chain)` or in case of error the pendingRequests must be answered.

Reply:
In addition based on this #443 - app should wait for a "connected" state in order to start sending RPC messages. Substrate-connect imo should not be a pending-message-holder

Reply:
If we track the app's "connected" state after #443 then suddenly we're tracking and communicating 2 different types of "connected"ness and complicating the protocol to do so.  This pending request tracking is a backstop for a super-edge case so as not to lose requests.  In reality the only supported use case for the extension is by our own `ExtensionProvider` and we know that the `ExtensionProvider` is always going to send the `'spec'` message immediately.  Once #313 (healthchecking) is implemented, we know that polkdotjs API will not make any RPC requests on behalf of the app until we know the chain in the background has stopped syncing.  However its quite possible that the healthchecker will send an 'RPC' message before the `addChain` completed in the background, this `pendingRequest` mechanism will simply hold onto those requests and respond to them (with a slight delay).  

> The flow below does not guarantee the deliverance of the messages (for eample if (!chain) or in case of error the pendingRequests must be answered.

I think that #442 and #439 cover this concern.  When those are implemented we will 1) send proper errors back for problems adding chains and 2) reject all the promises for the cases where the app has made some requests and, before they complete, the app receives a 'disconnect' or 'error' message back.

Reply:
I agree with the approach above based on your comment @raoulmillais . I think we cover 99.5% of the scenarios and there are extremely edge cases that we may have not thought off - or met yet.